sidechain/consensus/aura/src/peer_block_sync.rs

```diff
@@ -0,0 +1,214 @@
+}
+
+/// Sidechain peer block sync implementation.
+pub struct PeerBlockSync<
+	ParentchainBlock,
+	SignedSidechainBlock,
+	BlockImporter,
+	BlockProductionSuspender,
+> {
+	importer: Arc<BlockImporter>,
+	block_production_suspender: Arc<BlockProductionSuspender>,
+	_phantom: PhantomData<(ParentchainBlock, SignedSidechainBlock)>,
+}
+
+impl<ParentchainBlock, SignedSidechainBlock, BlockImporter, BlockProductionSuspender>
+	PeerBlockSync<ParentchainBlock, SignedSidechainBlock, BlockImporter, BlockProductionSuspender>
+where
+	ParentchainBlock: ParentchainBlockTrait,
+	SignedSidechainBlock: SignedSidechainBlockTrait,
+	BlockImporter: BlockImport<ParentchainBlock, SignedSidechainBlock>,
+	BlockProductionSuspender: SuspendBlockProduction + IsBlockProductionSuspended,
+{
+	pub fn new(
+		importer: Arc<BlockImporter>,
+		block_production_suspender: Arc<BlockProductionSuspender>,
+	) -> Self {
+		PeerBlockSync { importer, block_production_suspender, _phantom: Default::default() }
+	}
+
+	fn fetch_and_import_blocks_from_peer(
+		&self,
+		_last_known_block_hash: BlockHash,
+		last_imported_parentchain_header: &ParentchainBlock::Header,
+	) -> Result<()> {
+		// TODO fetch blocks from peer
+
+		let blocks_to_import: Vec<SignedSidechainBlock> = Vec::new();
+		for block_to_import in blocks_to_import {
+			self.importer.import_block(block_to_import, last_imported_parentchain_header)?;
+		}
+
+		Ok(())
+	}
+}
+
+impl<ParentchainBlock, SignedSidechainBlock, BlockImporter, BlockProductionSuspender>
+	SyncBlockFromPeer<ParentchainBlock::Header, SignedSidechainBlock>
+	for PeerBlockSync<ParentchainBlock, SignedSidechainBlock, BlockImporter, BlockProductionSuspender>
+where
+	ParentchainBlock: ParentchainBlockTrait,
+	SignedSidechainBlock: SignedSidechainBlockTrait,
+	BlockImporter: BlockImport<ParentchainBlock, SignedSidechainBlock>,
+	BlockProductionSuspender: SuspendBlockProduction + IsBlockProductionSuspended,
+{
+	fn attempt_block_sync(
+		&self,
+		sidechain_block: SignedSidechainBlock,
+		last_imported_parentchain_header: &ParentchainBlock::Header,
+	) -> Result<()> {
+		// In case block production is suspended, we don't import any blocks.
+		// In the future we might want to cache the blocks here, so they can be imported later.
+		if self.block_production_suspender.is_suspended().unwrap_or_default() {
+			warn!("Sidechain block won't be imported, since block production is suspended");
+			return Ok(())
+		}
+
+		// Attempt to import the block - in case we encounter an ancestry error, we go into
+		// peer fetching mode to fetch sidechain blocks from a peer and import those first.
+		match self.importer.import_block(sidechain_block, last_imported_parentchain_header) {
+			Err(e) => match e {
+				Error::BlockAncestryMismatch(_block_number, block_hash, _) => {
+					warn!("Got ancestry mismatch error upon block import. Attempting to fetch missing blocks from peer");
+
+					// TODO need a 'finally' (or on-drop) here for the production suspension,
```

Comment:
Actually, I think you have solved this TODO already be not returning upon an error in `fetch_and_import_blocks_from_peer`. The block production is resumed either way, which is perfect in my eyes. Because if somthing has failed and we still have a block mismatch, we will end up here again, so nothing is lost.

Reply:
I think the comment is about something different, but I don't grasp it 100% either.

Reply:
I can update the comment. It is meant for the case when we add more logic in between the `suspend` and `resume` parts, or have changes that introduce early returns. The way it is now is quite brittle, because we have to make sure that `resume` is always called (i.e. no `?` or panic in between). With a `finally`-like mechanism we could ensure that `resume` is always called.